pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -194,12 +196,42 @@ func receiveThread(con *XdsConnection, reqChannel chan *xdsapi.DiscoveryRequest,
 
 // StreamAggregatedResources implements the ADS interface.
 func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
-	peerInfo, ok := peer.FromContext(stream.Context())
+	ctx := stream.Context()
+	peerInfo, ok := peer.FromContext(ctx)
 	peerAddr := "0.0.0.0"
 	if ok {
 		peerAddr = peerInfo.Addr.String()
 	}
 
+	// Authenticate - currently just checks that request has a certificate signed with the our key.
+	// Protected by flag to avoid breaking upgrades - should be enabled in multi-cluster/meshexpansion where
+	// XDS is exposed.
+	if s.Authenticators != nil && len(s.Authenticators) > 0 {
+		if err := credentials.CheckSecurityLevel(ctx, credentials.PrivacyAndIntegrity); err == nil {
+			var authenticatedID *authenticate.Caller
+			for _, authn := range s.Authenticators {
+				u, err := authn.Authenticate(ctx)
+				if u != nil && err == nil {
+					authenticatedID = u
+				}
+			}
+
+			if authenticatedID == nil {
+				adsLog.Errora("Failed to authenticate client ", peerInfo)
+				return errors.New("authentication failure")
+			}
+
+			adsLog.Infoa("Authenticated XDS: ", peerInfo, " ", authenticatedID.AuthSource, " ", authenticatedID.Identities)
+		} else {
+			// TODO: add a flag to prevent unauthenticated requests ( 15010 )
+			// request not over TLS ( on the insecure port
+			adsLog.Infoa("Unauthenticated XDS: ", peerInfo)
+		}
+
+	}
+
+	// TODO: We should validate that the namespace in the cert matches the claimed namespace in metadata.
```

Comment:
nit: put todo inside auth

Reply:
I moved it where the implementation should be - next to the code getting the metadata.